File: bitset/bitset.go
package bitset

type (
	BitSet struct {
		bits []uint64
		size int
	}
)

func New(size int) *BitSet {
	return &BitSet{
		bits: make([]uint64, (size+63)/64),
		size: 0,
	}
}

func (s *BitSet) Reset() {
	for i := range s.bits {
		s.bits[i] = 0
	}
	s.size = 0
}

func (s *BitSet) Size() int {
	return s.size
}

func (s *BitSet) Contains(i int) bool {
	a := i / 64
	mask := uint64(1) << (i % 64)
	return (s.bits[a] & mask) != 0
}

func (s *BitSet) Set(i int, b bool) bool {
	a := i / 64
	mask := uint64(1) << (i % 64)
	if b {
		diff := (s.bits[a] & mask) == 0
		s.bits[a] |= mask
		if diff {
			s.size++
		}
		return diff
	} else {
		diff := (s.bits[a] & mask) != 0
		s.bits[a] &^= mask
		if diff {
			s.size--
		}
		return diff
	}
}

func (s *BitSet) Insert(i int) bool {
	a := i / 64
	mask := uint64(1) << (i % 64)
	diff := (s.bits[a] & mask) == 0
	s.bits[a] |= mask
	if diff {
		s.size++
	}
	return diff
}

func (s *BitSet) Remove(i int) bool {
	a := i / 64
	mask := uint64(1) << (i % 64)
	diff := (s.bits[a] & mask) != 0
	s.bits[a] &^= mask
	if diff {
		s.size--
	}
	return diff
}

File: src/bitset.rs
pub struct BitSet {
    bits: Vec<usize>,
    size: usize,
}

impl BitSet {
    pub fn new(size: usize) -> Self {
        Self {
            bits: vec![0; (size + (usize::BITS as usize - 1)) / usize::BITS as usize],
            size: 0,
        }
    }

    pub fn len(&self) -> usize {
        self.size
    }

    pub fn reset(&mut self) {
        for i in self.bits.iter_mut() {
            *i = 0;
        }
        self.size = 0;
    }

    pub fn contains(&self, i: usize) -> bool {
        let a = i / usize::BITS as usize;
        let b = i % usize::BITS as usize;
        let mask = 1 << b;
        self.bits[a] & mask != 0
    }

    pub fn insert(&mut self, i: usize) -> bool {
        let a = i / usize::BITS as usize;
        let mask = 1 << (i % usize::BITS as usize);
        let diff = (self.bits[a] & mask) == 0;
        self.bits[a] |= mask;
        if diff {
            self.size += 1;
        }
        diff
    }

    pub fn remove(&mut self, i: usize) -> bool {
        let a = i / usize::BITS as usize;
        let mask = 1 << (i % usize::BITS as usize);
        let diff = (self.bits[a] & mask) != 0;
        self.bits[a] &= !mask;
        if diff {
            self.size -= 1;
        }
        diff
    }
}
